client/shared/src/search/query/patternMatcher.ts

```diff
@@ -0,0 +1,408 @@
+ * property types from each union member, or 'undefined' if the property doesn't
+ * exist.
+ * Normally it's not possible to access a property on a union type that doesn't
+ * exist in all of the members but because we operate on a generic type and we
+ * extract all property names via 'UnionKeys', we can get away with it. However,
+ * doing just '{[K in UnionKeys<T>]: T[K]}' will result in 'unknown' if one of
+ * the members doesn't have the property.
+ * With 'UnionValues', we get '... | undefined' instead.
+ */
+type UnionValues<Value, K> = Value extends any ? (K extends keyof Value ? Value[K] : undefined) : never
+
+/**
+ * A ObjectPattern matches the input if all pattern properties (which are a
+ * subset of the input properties) match.
+ *
+ * Example: {foo: x => x > 0} matches {foo: 42}
+ *
+ * If the input type is a union of objects, the properties of all members are
+ * merged into into a single object. The type of every property is the union of
+ * the respective property types, including 'undefined' if one of the objects in
+ * the union doesn't have this property.
+ *
+ * Example:
+ * Input:
+ *     {t: 'foo', a: string, b: number} | {t: 'bar', a: number}
+ *
+ * Output:
+ *     ObjectPattern<Input> = {
+ *       t?: PatternFunction<'foo'|'bar'> | WrapperPattern<'foo'|'bar'> | 'foo' | 'bar',
+ *       a?: PatternFunction<string|number> | WrapperPattern<string|number> | string | number
+ *       b?: PatternFunction<string|undefined> | WrapperPattern<string|undefined> | string
+ *     }
+ */
+type ObjectPattern<Value, Data> =
+    // ObjectPattern is applied below with ObjectPattern<ObjectMembers<Value>>. The
+    // check against 'never' prevents empty unions.
+    [Value] extends [never]
+        ? never
+        : { [K in UnionKeys<Value>]?: PatternOf<UnionValues<Value, K>, Data> } & Partial<WrapperPattern<Value, Data>>
+
+/**
+ * A WrapperPattern makes it possible to add data annotations to patterns that
+ * are not object patterns. In the following example, it wouldn't be possible to
+ * attach additional data to the different values of "type":
+ *
+ * {type: oneOf("script", "module")}
+ *
+ * We could write
+ *
+ * {type: oneOf("script", "module"), $: {module: value => value.type === "module"}
+ *
+ * but in order to avoid precedural logic as much as possible, WrapperPattern
+ * allows us to write
+ *
+ * {type: oneOf("script", {_: "module", $: {module: true}})}
+ */
+interface WrapperPattern<Value, Data> extends DataAnnotation<Value, Data> {
+    _: PatternFunction<Value, Data> | PrimitivePattern<Value>
+}
+
+/**
+ * A PatternFunction accepts the value to match against, the current match
+ * context (which also holds the extracted data) and the internal match function
+ * for convenience. See the combinatorial helper functions below for examples.
+ */
+export type PatternFunction<Value, Data> = (
+    value: Value,
+    context: MatchContext<Data>,
+    matchfn: typeof matches
```

Comment:
Another nit naming comment -- but this time suggesting a longer name 😆 ! `fn` is implied by the time.

```suggestion
    match: typeof matches
```

Reply:
FWIW, I've also considered to expose the match function as part of `context`.